### PR TITLE
[MLv2] Handle coalesce type after conversion

### LINKED ISSUE #33680: Can't aggregate with a custom column that uses coalesce
### Describe the bug

When you have a custom column that is a number created by a coalesce in the custom column, you can not choose it in the Summarize section for math aggregations. 

### To Reproduce

1. Go to Sample Data
2. Click on Orders
3.  Join to the People table
4.  Create a custom column, "Zero":
image

5. Create another custom column, "Total with Zeros": 
image

6. Go to the Summarize section, choose "Sum of" and attempt to pick "Total with Zeros"

### Expected behavior

It should allow you to pick a custom column that is a number created with a coalesce function

### Logs

_No response_

### Information about your Metabase installation

```JSON
It does not work on master or 47

It does work in 1.45.3.1 and 1.46.6.1-latest patch
```

### Severity

It's pretty annoying.  Also not sure what happens if you upgrade - will these reports stop working?

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by snoe**
Resolves #33680 

I'm not sure if this will be too slow, but missing types on field refs will cause coalesce to lose the necessary hints to figure it out. Is there a better place to do this?

### CODE REVIEW

File: test/metabase/lib/aggregation_test.cljc [hunk 1]
```diff
@@ -766,3 +766,28 @@
               query))
       (is (=? [:field {:lib/uuid string?, :base-type :type/Float} "avg"]
               (lib/ref expr-metadata))))))
+
+(deftest ^:parallel aggregate-by-coalesce-test
+  (testing "Converted query returns same columns as built query"
```
Comment (metamben) [inline test/metabase/lib/aggregation_test.cljc:771]:
```suggestion
  (testing "Converted query returns same columns as built query (#33680)"
```

**Review by metamben — APPROVED** [review]
I think this should work, but indeed doesn't seem to be the best place for the fix.
Ideally, MLv2 queries should not need this fix, which means it could be a task for the conversion to pMBQL.

File: src/metabase/lib/schema/expression.cljc [hunk 2]
```diff
@@ -139,7 +139,12 @@
 
 (def equality-comparable-types
   "Set of base types that can be campared with equality."
-   #{:type/Boolean :type/Text :type/Number :type/Temporal})
+   #{:type/Boolean :type/Text :type/Number :type/Temporal :type/IPAddress :type/MongoBSONID :type/Array})
```
Comment (snoe) [inline src/metabase/lib/schema/expression.cljc:142]: by filling out types for legacy queries, it unearthed some missing types

**Review by metamben — APPROVED** [review]
🏅

**Pull request merged**